### DISCUSSION TITLE
Differnt width strings

### DISCUSSION THREAD

**Discussion by marsfan** [NationalSecurityAgency/ghidra#4135] [Q&A]
I am working on reverse engineering a system that has the string stored in 16 bit little endian form, but still ASCII. 
 (I learned that this is how it is set up with [this question I asked on stack overflow](https://reverseengineering.stackexchange.com/questions/30167/contents-of-spi-nor-flash-appears-scrambled/30168?noredirect=1#comment45430_30168))

For example, in this snippet, the second word in the middle line is `FAHRENHEIT`.
```
00070f10  20 54 52 50 4a 4f 43 45  49 54 4e 4f 41 20 54 4c  | TRPJOCEITNOA TL|
00070f20  57 2f 54 49 20 48 41 46  52 48 4e 45 45 48 54 49  |W/TI HAFRHNEEHTI|
00070f30  20 20 45 43 53 4c 55 49  20 53 52 50 4a 4f 00 00  |  ECSLUI SRPJO..|
00070f40  00 00 20 20 55 43 52 52  4e 45 20 54 20 20 52 20  |..  UCRRNE T  R |
```

Is there a way to tell ghidra that the string is encoded in this way to make displaying the data easier?

**Comment by dev747368** [comment]
This type of string data is something we're aware of but haven't fully supported yet.  (Ghidra does handle some architectures that are similar in that they embed a 8 bit char value in a 16 bit word.  The string decoding logic takes its cue from the dataOrganization's char size, as specified by the cspec)

One possible workaround is to assign each string a 'translated' value that will be shown in the UI instead of the value as read from memory.  You could tweak the logic of SampleStringTranslationPlugin.java or TranslateStringScript.java to implement the unscrambling.

**Reply by astrelsky** [reply]
This is unfortunately one of the most underused and unappreciated features.

Oh I've got some obfuscated text, let me go write a script to add **comments** in the listing and decompiler so I know what they are. /end rant

**Reply by marsfan** [reply]
That's frustrating that it is not really there yet, but at least there is the translation as a workaround. Might be trickier for me since the storage can even move around the null terminators, but I can probably make it work.

**Reply by astrelsky** [reply]
> That's frustrating that it is not really there yet, but at least there is the translation as a workaround. Might be trickier for me since the storage can even move around the null terminators, but I can probably make it work. 

It is fully there, it's just not widely known about or used as much as it should be.

You will be able to accomplish this using the translation feature. You can specify the value in the gui by creating the string and then going to the translation option in the right click menu. You can automate it using a script as dev747368 mentioned above. 

You could go further and write a small plugin that would add an appropriate `DataType` which would automatically do  it when creating the type. I forget the details though.

**Reply by astrelsky** [reply]
Stupid GitHub app won't let me edit.

Anyway, if this is a one off thing where you will only need it for analyzing this program it's definitely easier to just write a small script to use the `currentAddress` or `currentSelection` to create the string, apply the correct value and just bind the script to a key binding.
